src/backend/cuda/kernel/random.hpp

```diff
@@ -114,6 +255,301 @@ namespace kernel
         states[id] = state;
     }
 
+    __device__ static float normalizeToFloat(const uint& num)
+    {
+        return float(num)/uintmaxfloat;
+    }
+
+    __device__ static double normalizeToDouble(const uintl& num)
+    {
+        return double(num)/uintlmaxdouble;
+    }
+
+    __device__ static void writeOut(uint *out, const unsigned index,
+            const PhiloxCounter<uint, 4>& counter)
+    {
+        out[index]                  = counter.v[0];
+        out[index + blockDim.x]     = counter.v[1];
+        out[index + 2*blockDim.x]   = counter.v[2];
+        out[index + 3*blockDim.x]   = counter.v[3];
+    }
+
+    __device__ static void writeOut(uintl *out, const unsigned &index,
+            const PhiloxCounter<uint, 4>& counter)
+    {
+        out[index]              = (uintl(counter.v[0])<<32) | uintl(counter.v[1]);
+        out[index + blockDim.x] = (uintl(counter.v[2])<<32) | uintl(counter.v[3]);
+    }
+
+    __device__ static void writeOut(int *out, const unsigned index,
+            const PhiloxCounter<uint, 4>& counter)
+    {
+        out[index]                  = counter.v[0];
+        out[index + blockDim.x]     = counter.v[1];
+        out[index + 2*blockDim.x]   = counter.v[2];
+        out[index + 3*blockDim.x]   = counter.v[3];
+    }
+
+    __device__ static void writeOut(intl *out, const unsigned &index,
+            const PhiloxCounter<uint, 4>& counter)
+    {
+        out[index]              = (uintl(counter.v[0])<<32) | uintl(counter.v[1]);
+        out[index + blockDim.x] = (uintl(counter.v[2])<<32) | uintl(counter.v[3]);
+    }
+
```

Comment:
@kaatish I think it might be better to macro-ize these device kernels and reduce the amount of code. rather than performing that hack.